### ISSUE TITLE
HaskellCompletionCacheLoader needs read lock

### ISSUE THREAD

**Issue by carymrobbins** [carymrobbins/intellij-haskforce#271]
```
Read access is allowed from event dispatch thread or inside read-action only (see com.intellij.openapi.application.Application.runReadAction())
Details: Current thread: Thread[ApplicationImpl pooled thread 378,4,main] 2042245425
; dispatch thread: false; isDispatchThread(): false
SystemEventQueueThread: Thread[AWT-EventQueue-0 2016.1#IU-145.596.7 IDEA, eap:true, os:Linux 3.16.0-57-generic, java-version:JetBrains s.r.o 1.8.0_76-release-b18,6,main] 10089927
java.lang.Throwable
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:138)
	at com.intellij.openapi.application.impl.ApplicationImpl.assertReadAccessAllowed(ApplicationImpl.java:1054)
	at com.intellij.psi.impl.file.impl.FileManagerImpl.findFile(FileManagerImpl.java:393)
	at com.intellij.psi.impl.PsiManagerImpl.findFile(PsiManagerImpl.java:184)
	at com.haskforce.codeInsight.HaskellCompletionCacheLoader$MyHandler.fileContentLoaded(HaskellCompletionCacheLoader.scala:47)
	at sun.reflect.GeneratedMethodAccessor26.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.util.messages.impl.MessageBusConnectionImpl.deliverMessage(MessageBusConnectionImpl.java:117)
	at com.intellij.util.messages.impl.MessageBusImpl.doPumpMessages(MessageBusImpl.java:372)
	at com.intellij.util.messages.impl.MessageBusImpl.pumpMessages(MessageBusImpl.java:359)
	at com.intellij.util.messages.impl.MessageBusImpl.pumpMessages(MessageBusImpl.java:345)
	at com.intellij.util.messages.impl.MessageBusImpl.sendMessage(MessageBusImpl.java:336)
	at com.intellij.util.messages.impl.MessageBusImpl.access$200(MessageBusImpl.java:42)
	at com.intellij.util.messages.impl.MessageBusImpl$2.invoke(MessageBusImpl.java:226)
	at com.sun.proxy.$Proxy95.repositoryChanged(Unknown Source)
	at git4idea.repo.GitRepositoryImpl$1.run(GitRepositoryImpl.java:225)
	at com.intellij.openapi.application.impl.ApplicationImpl$8.run(ApplicationImpl.java:369)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by carymrobbins** [comment]
This should be resolved in the next release: **0.3-beta.35**

**Comment by pnaranja** [comment]
Hi, I'm getting the same exception on:
Windows 10
Java 1.8.0_92
IntelliJ IDEA 2016.2 Community Edition
intellij-haskforce 0.3-beta.35

Let me know if you need more info and/or open a new issue. 


Read access is allowed from event dispatch thread or inside read-action only (see com.intellij.openapi.application.Application.runReadAction())
Details: Current thread: Thread[ApplicationImpl pooled thread 102,4,main] 826858301
; dispatch thread: false; isDispatchThread(): false
SystemEventQueueThread: Thread[AWT-EventQueue-0 2016.2#IC-162.1121.32 IDEA, eap:false, os:Windows 10 10.0, java-version:Oracle Corporation 1.8.0_92-b14,6,main] 298453087
java.lang.Throwable
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:138)
	at com.intellij.openapi.application.impl.ApplicationImpl.assertReadAccessAllowed(ApplicationImpl.java:950)
	at com.intellij.psi.impl.file.impl.FileManagerImpl.findFile(FileManagerImpl.java:385)
	at com.intellij.psi.impl.PsiManagerImpl.findFile(PsiManagerImpl.java:185)
	at com.haskforce.cabal.completion.CabalFileFinder$.com$haskforce$cabal$completion$CabalFileFinder$$getPsi(CabalFileFinder.scala:36)
	at com.haskforce.cabal.completion.CabalFileFinder$$anonfun$psiForFile$1.apply(CabalFileFinder.scala:32)
	at com.haskforce.cabal.completion.CabalFileFinder$$anonfun$psiForFile$1.apply(CabalFileFinder.scala:32)
	at scala.Option.flatMap(Option.scala:171)
	at com.haskforce.cabal.completion.CabalFileFinder$.psiForFile(CabalFileFinder.scala:32)
	at com.haskforce.cabal.query.BuildInfoUtil$.getBuildInfo(BuildInfoUtil.scala:12)
	at com.haskforce.cabal.query.BuildInfoUtil.getBuildInfo(BuildInfoUtil.scala)
	at com.haskforce.highlighting.annotation.external.HLint.getParamsFromCabal(HLint.java:118)
	at com.haskforce.highlighting.annotation.external.HLint.getParams(HLint.java:113)
	at com.haskforce.highlighting.annotation.external.HLint.access$100(HLint.java:45)
	at com.haskforce.highlighting.annotation.external.HLint$3.apply(HLint.java:91)
	at com.haskforce.highlighting.annotation.external.HLint$3.apply(HLint.java:87)
	at scala.util.Either.fold(Either.scala:99)
	at com.haskforce.highlighting.annotation.external.HLint.parseProblems(HLint.java:80)
	at com.haskforce.highlighting.annotation.external.HLint.lint(HLint.java:58)
	at com.haskforce.highlighting.annotation.external.impl.HLintProblemsProvider.getProblems(HLintProblemsProvider.scala:24)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$$anonfun$buildForFile$1$$anonfun$apply$1.apply(HaskellExternalAnnotator.scala:94)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$$anonfun$buildForFile$1$$anonfun$apply$1.apply(HaskellExternalAnnotator.scala:94)
	at scala.Option.map(Option.scala:146)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$$anonfun$buildForFile$1.apply(HaskellExternalAnnotator.scala:94)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$$anonfun$buildForFile$1.apply(HaskellExternalAnnotator.scala:94)
	at scala.collection.immutable.List.flatMap(List.scala:327)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator$State$.buildForFile(HaskellExternalAnnotator.scala:94)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator.doAnnotate(HaskellExternalAnnotator.scala:44)
	at com.haskforce.highlighting.annotation.external.HaskellExternalAnnotator.doAnnotate(HaskellExternalAnnotator.scala:19)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.doAnnotate(ExternalToolPass.java:212)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass.access$500(ExternalToolPass.java:44)
	at com.intellij.codeInsight.daemon.impl.ExternalToolPass$1.run(ExternalToolPass.java:160)
	at com.intellij.util.ui.update.MergingUpdateQueue.execute(MergingUpdateQueue.java:320)
	at com.intellij.util.ui.update.MergingUpdateQueue.execute(MergingUpdateQueue.java:310)
	at com.intellij.util.ui.update.MergingUpdateQueue.lambda$flush$1(MergingUpdateQueue.java:260)
	at com.intellij.util.ui.update.MergingUpdateQueue.flush(MergingUpdateQueue.java:274)
	at com.intellij.util.ui.update.MergingUpdateQueue.run(MergingUpdateQueue.java:229)
	at com.intellij.util.concurrency.QueueProcessor.runSafely(QueueProcessor.java:222)
	at com.intellij.util.Alarm$Request$1.run(Alarm.java:377)
	at com.intellij.util.Alarm$Request.run(Alarm.java:388)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at com.intellij.util.concurrency.SchedulingWrapper$MyScheduledFutureTask.run(SchedulingWrapper.java:227)
	at com.intellij.util.concurrency.BoundedTaskExecutor.runFirstTaskThenPollAndRunRest(BoundedTaskExecutor.java:178)
	at com.intellij.util.concurrency.BoundedTaskExecutor.access$000(BoundedTaskExecutor.java:40)
	at com.intellij.util.concurrency.BoundedTaskExecutor$2.run(BoundedTaskExecutor.java:197)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

**Comment by carymrobbins** [comment]
Thanks for the report @pnaranja, this should be resolved in the next release, **0.3-beta.36**
